ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V1.hs

```diff
@@ -746,7 +753,18 @@ newForkerByRollback ::
   Word64 ->
   m (Either GetForkerError (Forker m l blk))
 newForkerByRollback h rr n = getEnv h $ \ldbEnv -> do
-  withReadLock (ldbLock ldbEnv) (acquireAtTarget ldbEnv (Left n) >>= traverse (newForker h ldbEnv rr))
+  tv <- newTVarIO (pure ())
+  void $ allocate rr
+    (\_ -> atomically $ do
+        -- Populate the tvar with the releasing action. Once we create the
+        -- forker the release will be done via `forkerClose` if an exception
+        -- makes us deallocate the forker
+        writeTVar tv (atomically $ unsafeReleaseReadAccess (ldbLock ldbEnv))
+        -- Acquire the read access
+        unsafeAcquireReadAccess (ldbLock ldbEnv)
+    )
+    (\_ -> join $ readTVarIO tv)
+  runReadLocked (acquireAtTarget ldbEnv (Left n) >>= traverse (newForker h ldbEnv tv rr))
```

Comment:
The code between `newForkerByRollback` and `newForkerAtTarget` should probably be shared somehow